docker.go

```diff
@@ -174,41 +187,177 @@ func runCommands(commands []string) error {
+			folder = temp
+		} else {
+			if ib.Instructions != "" {
+				out = must(ioutil.TempFile(ib.Dir(), "DockerFile")).(*os.File)
+				temp = out.Name()
+				dockerFile = temp
+			}
+			folder = ib.Dir()
+		}
+
+		if out != nil {
+			ib.Instructions = fmt.Sprintf("FROM %s\n%s\n", name, ib.Instructions)
+			must(fmt.Fprintf(out, ib.Instructions))
+			must(out.Close())
+		}
+
+		if temp != "" {
+			// A temporary file of folder has been created, we register functions to ensure proper cleanup
+			cleanup := func() { os.Remove(temp) }
+			defer cleanup()
+			c := make(chan os.Signal, 1)
+			signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+			go func() {
+				<-c
+				Println("\nRemoving file", dockerFile)
+				cleanup()
+				panic(errorString("Execution interrupted by user: %v", c))
+			}()
+		}
+
+		name = name + "-" + ib.Tag()
+		if refresh || getActualImageVersionInternal(name) == "" {
+			args := []string{"build", ".", "--quiet", "--force-rm"}
+			if i == 0 && refresh {
+				args = append(args, "--pull")
+			}
+			if dockerFile != "" {
+				args = append(args, "--file")
+				args = append(args, filepath.Base(dockerFile))
+			}
+
+			args = append(args, "--tag", name)
+			buildCmd := exec.Command("docker", args...)
+
+			debugPrint("%s", strings.Join(buildCmd.Args, " "))
+			if ib.Instructions != "" {
+				debugPrint("%s", ib.Instructions)
+			}
+			buildCmd.Stderr = os.Stderr
+			buildCmd.Dir = folder
+			must(buildCmd.Output())
+			prune()
+		}
 	}
 
-	if config.ImageBuildFolder == "" {
-		config.ImageBuildFolder = "."
+	return
+}
+
+func prune(images ...string) {
+	if len(images) > 0 {
+		current := fmt.Sprintf(">=%s", GetActualImageVersion())
+		cli, ctx := getDockerClient()
+		for _, image := range images {
+			filters := filters.NewArgs()
+			filters.Add("reference", image)
+			if images, err := cli.ImageList(ctx, types.ImageListOptions{Filters: filters}); err == nil {
+				for _, image := range images {
+					actual := getActualImageVersionFromImageID(image.ID)
+					if actual == "" {
+						for _, tag := range image.RepoTags {
+							matches, _ := utils.MultiMatch(tag, reVersion)
+							if version := matches["version"]; version != "" {
+								if len(version) > len(actual) {
+									actual = version
+								}
+							}
+						}
+					}
+					upToDate, err := CheckVersionRange(actual, current)
+					if err != nil {
+						ErrPrintln("Check version for %s vs%s: %v", actual, current, err)
+					} else if !upToDate {
+						for _, tag := range image.RepoTags {
+							deleteImage(tag)
+						}
+					}
+				}
+			}
+		}
 	}
 
-	var dockerFile string
-	if config.ImageBuild != "" {
-		out := Must(ioutil.TempFile(config.ImageBuildFolder, "DockerFile")).(*os.File)
-		Must(fmt.Fprintf(out, "FROM %s \n%s", config.GetImageName(), config.ImageBuild))
-		Must(out.Close())
-		defer os.Remove(out.Name())
-		dockerFile = out.Name()
+	pruneCmd := exec.Command("docker", "system", "prune", "-f")
```

Comment:
You use the docker client in other places. Have you tried using it here?

Reply:
I considered it, but was too lazy 😜.